post.md

```diff
@@ -0,0 +1,166 @@
+  const validation = schema.safeParse({
+    name: params.get("name"),
+    email: params.get("email"),
+  });
+
+  if (validation.success) {
+    // save the data, send an email, etc.
+    redirect("/");
+  } else {
+    return {
+      errors: validation.error.issues,
+    };
+  }
+}
+```
+
+## Contact Page
+
+In our contact page, we import the server action and pass it to the `Form` component. The `ContactPage` is a server component while the `Form` component is a client component. We need to use the `useFormState` hook to handle validation errors coming from the server. The `useFormState` hook can only be used in client components.
+
+```tsx
+// app/contact/page.tsx
+import contactAction from "./server-action";
+import Form from "./form";
+
+export default async function ContactPage() {
+  return (
+    <>
+      <h1 className="text-2xl font-bold mb-3">Contact Us</h1>
+      <Form action={contactAction}></Form>
+    </>
+  );
+}
+```
+
+## Form Component
+
+In our form component, we use the `useFormState` hook to handle validation errors. We match the `path` node of each error object to the corresponding input field and extract all `message` nodes to build the error message for each field.
+
+```tsx
+// app/contact/form.tsx
+"use client";
+
+import { useFormState, useFormStatus } from "react-dom";
+import type { ZodIssue } from "zod";
+
+type Props = {
+  action: (
+    _prevState: any,
+    params: FormData
+  ) => Promise<{ errors: ZodIssue[] }>;
+};
+
+export default function Form({ action }: Props) {
+  const [state, formAction] = useFormState(action, { errors: [] });
+  console.log(state);
+  const nameErrors = findErrors("name", state.errors);
+  const emailErrors = findErrors("email", state.errors);
+
+  return (
+    <form action={formAction}>
+      <div>
+        <label htmlFor="name">
+          <span>Name</span>
+          <input id="name" type="text" name="name" />
+        </label>
+        <ErrorMessages errors={nameErrors} />
+      </div>
+
+      <div>
+        <label htmlFor="email">
+          <span>Email</span>
+          <input id="email" type="text" name="email" />
+        </label>
+        <ErrorMessages errors={emailErrors} />
+      </div>
+
+      <button type="submit">Submit</button>
+    </form>
+  );
+}
+
+const ErrorMessages = ({ errors }: { errors: string[] }) => {
+  if (errors.length === 0) return null;
+
+  const text = errors.join(", ");
+
+  return <div className="text-red-600 peer">{text}</div>;
+};
+
+const findErrors = (fieldName: string, errors: ZodIssue[]) => {
+  return errors
+    .filter((item) => {
+      return item.path.includes(fieldName);
+    })
+    .map((item) => item.message);
+};
+```
+
+In conclusion, Zod provides a powerful and flexible way to handle form validation in a Next.js application. By combining it with server-side actions and the `useFormState` hook, we can create robust forms with clear and specific error messages.
```

Comment:
A few things here:
- I suggest adding a section title. Something like, "Wrapping it Up"
- Let's flesh out more how the "why" is accomplished at this point